database/manager/manager.go

```diff
@@ -0,0 +1,252 @@
+// underlying database.
+func (m *manager) wrapManager(wrap func(db *VersionedDatabase) (*VersionedDatabase, error)) (*manager, error) {
+	databases := make([]*VersionedDatabase, 0, len(m.databases))
+	for _, db := range m.databases {
+		wrappedDB, err := wrap(db)
+		if err != nil {
+			// ignore additional errors in favor of returning the original error
+			_ = closeDBs(databases)
+			return nil, err
+		}
+		databases = append(databases, wrappedDB)
+	}
+	return &manager{databases: databases}, nil
+}
+
+// NewDefaultMemDBManager returns a database manager with a single memory db instance
+// with a default version of v1.0.0
+func NewDefaultMemDBManager() Manager {
+	return &manager{
+		databases: []*VersionedDatabase{
+			{
+				Database: memdb.New(),
+				Version:  version.NewDefaultVersion(1, 0, 0),
+			},
+		},
+	}
+}
+
+// New creates a database manager at [filePath] by creating a database instance from each directory
+// with a version <= [currentVersion]
+func New(dbDirPath string, currentVersion version.Version) (Manager, error) {
+	parser := version.NewDefaultParser()
+
+	currentDBPath := path.Join(dbDirPath, currentVersion.String())
+	currentDB, err := leveldb.New(currentDBPath, 0, 0, 0)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create db at %s: %w", currentDBPath, err)
+	}
+
+	semDBs := []*VersionedDatabase{
+		{
+			Database: currentDB,
+			Version:  currentVersion,
+		},
+	}
+	err = filepath.Walk(dbDirPath, func(path string, info os.FileInfo, err error) error {
+		// Skip the root directory
+		if path == dbDirPath {
+			return nil
+		}
+
+		// The database directory should only contain database directories, no files.
+		if !info.IsDir() {
+			return fmt.Errorf("unexpectedly found non-directory at %s", path)
+		}
+		_, dbName := filepath.Split(path)
+		version, err := parser.Parse(dbName)
+		if err != nil {
+			return err
+		}
+
+		// If [version] is greater than or equal to the specified version
+		// skip over creating the new database to avoid creating the same db
+		// twice or creating a database with a version ahead of the desired one.
+		if cmp := version.Compare(currentVersion); cmp >= 0 {
+			return filepath.SkipDir
+		}
+
+		db, err := leveldb.New(path, 0, 0, 0)
+		if err != nil {
+			return fmt.Errorf("couldn't create db at %s: %w", path, err)
+		}
+
+		semDBs = append(semDBs, &VersionedDatabase{
+			Database: db,
+			Version:  version,
+		})
+
+		return filepath.SkipDir
+	})
+
+	SortDescending(semDBs)
+
+	m := &manager{
+		databases: semDBs,
+	}
+
+	// If an error occurred, close all of the opened databases
+	// and return the original error
+	if err != nil {
+		_ = m.Close()
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewPrefixDBManager creates a new manager with each database instance prefixed
+// by [prefix]
+func (m *manager) NewPrefixDBManager(prefix []byte) Manager {
+	m, _ = m.wrapManager(func(sdb *VersionedDatabase) (*VersionedDatabase, error) {
```

Comment:
I think that even if it's not _currently_ possible for `wrapManager` to return an error here, we should still check for it to future-proof the code

Reply:
`sDB` --> `vDB`

Reply:
Updated sDB -> vDB. I'd prefer not to make this change since it keeps the return type the same as before, and it's documented exactly why this will never return an error.